### PR TITLE
Fix timeout bug and enable Datadog tracing

### PR DESCRIPTION
**Pull request by newtrat**
## Description
The API is already using commit b25809e, so the only real change here is fixing Datadog tracing. However, I'm opening the PR so it's easy to CR and so we can merge all of it into master if we want.

Total changes:
- Fix a bug which could make Dataloader crash when a request timed out
- Allow a configurable Task module, so we can use `Datadog.Tracer` instead of the native `Task` within our repos.

## Motivation and Context
We want to trace dataloaded queries. Right now there's no connection between the API call which uses dataloader and the underlying Rails or Ecto calls from within the load functions.

## Testing
Test as part of the corresponding API PR, TheRealReal/api#1762

## Type of Changes

- [x] Bug fix _(a non-breaking change which fixes an issue)_
- [x] New feature _(a non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that changes existing functionality)_
- [ ] Bump dependencies
- [ ] Automated Testing / Missing tests
- [ ] Documentation

## Checklist:

- [ ] Yes, I have added tests to cover my changes.
- [ ] No, We don't have a test suite in this project yet.
- [ ] My change requires a change to the documentation.
- [ ] My change requires release strategy _(uncomment the Release strategy below)_

cc/ @TheRealReal/buyer-purchase

### CODE REVIEW

**Review by desmondgee — APPROVED** [review]
Thanks for finding out the task_module load order issue! LGTM.

File: lib/dataloader.ex [hunk 1]
```diff
@@ -210,7 +210,7 @@ defmodule Dataloader do
 
   @spec pending_batches?(t) :: boolean
   def pending_batches?(loader) do
-    Enum.any?(loader.sources, fn {_name, source} -> Source.pending_batches?(source) end)
+    Enum.any?(loader.sources, fn {_name, source} -> source != :timeout && Source.pending_batches?(source) end)
```
Comment (paradox460) [inline lib/dataloader.ex:213]: IMO a bit of pattern matching would be cleaner here

```suggestion
    Enum.any?(loader.sources, fn
      {_name, :timeout} -> false
      {_name, source} -> Source.pending_batches?(source)
    end)
```

**Review by paradox460 — APPROVED** [review]
1 comment, but its opinion so ¯\\\_(ツ)_/¯

**Review by landonmarder — APPROVED** [review]
💪 LGTM

wdyt about breaking this out into 2 separate releases?
1. re-adding tracing, verifying that it works with picking and feeds
2. adding the timeout

**Comment by newtrat** [comment]
@landonmarder IIRC the timeout fix is necessary for Dataloader to work correctly; otherwise the API will have an exception whenever the backend times out. The API's using the timeout fix already, since it's using [the second commit in this PR](https://github.com/TheRealReal/dataloader/commit/b25809e41fef06ee152ee27538291d31a3fcab3e)

**Comment by newtrat** [comment]
The API's been using this successfully for a while now, so I'll merge this PR in.

**Pull request merged**